### PR TITLE
feature/spaceship: Clause 22: Container Iterators

### PR DESCRIPTION
**Pull request by miscco**
This implements spaceship for the container iterators

### CODE REVIEW

File: stl/inc/vector [hunk 1]
```diff
@@ -2124,6 +2130,14 @@ public:
         return this->_Myptr == _Right._Myptr && this->_Myoff == _Right._Myoff;
     }
 
+#if _HAS_CXX20
+    _NODISCARD strong_ordering operator<=>(const _Vb_const_iterator& _Right) const noexcept {
+        if (const auto _CmpResult = this->_Myptr <=> _Right._Myptr; _CmpResult != 0) {
```
Comment (CaseyCarter) [inline stl/inc/vector:2135]: Note for other reviewers: `_Myptr` is a raw pointer, not a fancy pointer, so we can use `<=>` directly.

File: tests/std/tests/P1614R2_spaceship/test.cpp [hunk 2]
```diff
@@ -218,6 +239,15 @@ void ordering_test_cases() {
...
         static_assert((a0 <=> a2) > 0);
+
+        constexpr auto b1 = a1.begin();
+        constexpr auto b2 = a1.begin();
+        constexpr auto e1 = a1.end();
+
+        static_assert((b1 <=> b1) == 0);
+        static_assert((b1 <=> b2) == 0);
+        static_assert((b1 <=> e1) < 0);
+        static_assert((e1 <=> b1) > 0);
```
Comment (CaseyCarter) [inline tests/std/tests/P1614R2_spaceship/test.cpp:250]: These need to be non-`constexpr` variables and runtime `assert`s, per discussion on Discord. (Also 259-257.)

File: tests/std/tests/P1614R2_spaceship/test.cpp [hunk 3]
```diff
@@ -518,6 +581,9 @@ void ordering_test_cases() {
         const std::filesystem::path p3{R"(a/b/d)"};
 
         spaceship_test<std::strong_ordering>(p1, p2, p3);
+        spaceship_test<std::strong_ordering>(p1, p2, p3);
+        spaceship_test<std::strong_ordering>(p1, p2, p3);
```
Comment (CaseyCarter) [inline tests/std/tests/P1614R2_spaceship/test.cpp:585]: Why are we repeating ourselves repeating ourselves?

File: tests/std/tests/P1614R2_spaceship/test.cpp [hunk 4]
```diff
@@ -178,6 +178,27 @@ void unordered_containers_test(
...
+template <class Iter, class ConstIter>
+void ordered_iterator_test(const Iter& smaller, const Iter& smaller_equal, const Iter& larger,
+    const ConstIter& const_smaller, const ConstIter& const_smaller_equal, const ConstIter& const_larger) {
+    spaceship_test<std::strong_ordering>(smaller, smaller_equal, larger);
+    spaceship_test<std::strong_ordering>(const_smaller, const_smaller_equal, const_larger);
+    spaceship_test<std::strong_ordering>(const_smaller, smaller_equal, larger);
+}
+
+template <class Iter, class ConstIter>
+void unordered_iterator_test(const Iter& something, const Iter& something_equal, const Iter& different,
```
Comment (CaseyCarter) [inline tests/std/tests/P1614R2_spaceship/test.cpp:190]: I'm not sure these `unordered_iterator_test` cases belong here, or aren't redundant with coverage elsewhere, but I suppose they aren't hurting anything.

**Review by CaseyCarter — CHANGES_REQUESTED** [review]
What about `basic_string` and `basic_string_view`? (`span` is also notably covered by the Standard requirement, but we certainly have `<=>` for its iterators with test coverage elsewhere.)

File: tests/std/tests/P1614R2_spaceship/test.cpp (hunk 2 above)
Comment (miscco) [reply tests/std/tests/P1614R2_spaceship/test.cpp]: Removed as we test the non constexpr case below

File: tests/std/tests/P1614R2_spaceship/test.cpp (hunk 3 above)
Comment (miscco) [reply tests/std/tests/P1614R2_spaceship/test.cpp]: Ctrl CCC

**Comment by miscco** [comment]
> basic_string_view

Urgh char_traits and string literal overloads...

**Comment by CaseyCarter** [comment]
> > basic_string_view
> 
> Urgh char_traits and string literal overloads...

The iterator `<=>` doesn't care about these things. (And the range `<=>` is in another PR so not your problem.)

**Comment by miscco** [comment]
> > > basic_string_view
> > 
> > 
> > Urgh char_traits and string literal overloads...
> 
> The iterator `<=>` doesn't care about these things. (And the range `<=>` is in another PR so not your problem.)

I just pushed a commit for that ;) I like it when you agree that others should do the hard work

File: tests/std/tests/P1614R2_spaceship/test.cpp [hunk 5]
```diff
@@ -312,6 +295,20 @@ void ordering_test_cases() {
         ordered_containers_test(a1, a2, b1);
         ordered_iterator_test(a1.begin(), a1.begin(), a1.end(), a1.cbegin(), a1.cbegin(), a1.cend());
     }
+    { // string
+        std::string a1 = "aaa";
+        std::string a2 = "aaa";
+        std::string b1 = "bb";
+        // ordered_containers_test(a1, a2, b1);
```
Comment (CaseyCarter) [inline tests/std/tests/P1614R2_spaceship/test.cpp:302]:
```suggestion
        // ordered_containers_test(a1, a2, b1); // TRANSITION, GH-1635
```

File: tests/std/tests/P1614R2_spaceship/test.cpp [hunk 6]
```diff
@@ -312,6 +295,20 @@ void ordering_test_cases() {
...
+        std::string a2 = "aaa";
+        std::string b1 = "bb";
+        // ordered_containers_test(a1, a2, b1);
+        ordered_iterator_test(a1.begin(), a1.begin(), a1.end(), a1.cbegin(), a1.cbegin(), a1.cend());
+    }
+    { // string_view
+        std::string_view a1 = "aaa";
+        std::string_view a2 = "aaa";
+        std::string_view b1 = "bb";
+        // ordered_containers_test(a1, a2, b1);
```
Comment (CaseyCarter) [inline tests/std/tests/P1614R2_spaceship/test.cpp:309]:
```suggestion
        // ordered_containers_test(a1, a2, b1); // TRANSITION, GH-1635
```

**Review by CaseyCarter — APPROVED** [review]
I've made a small change to add `TRANSITION, GH-1635` to the two commented-out lines that would/could/should cover that work.

File: tests/std/tests/P1614R2_spaceship/test.cpp [hunk 7]
```diff
@@ -232,61 +254,86 @@ void ordering_test_cases() {
...
         std::vector<int> a2(3, 100);
         std::vector<int> b1(2, 200);
         ordered_containers_test(a1, a2, b1);
+        ordered_iterator_test(a1.begin(), a1.begin(), a1.end(), a1.cbegin(), a1.cbegin(), a1.cend());
+    }
+    { // string
+        std::string a1 = "aaa";
+        std::string a2 = "aaa";
+        std::string b1 = "bb";
+        // ordered_containers_test(a1, a2, b1); // TRANSITION, GH-1635
```
Comment (CaseyCarter) [inline tests/std/tests/P1614R2_spaceship/test.cpp:302]:
```suggestion
#if 0 // TRANSITION, GH-1635
        std::string a2 = "aaa";
        std::string b1 = "bb";
        ordered_containers_test(a1, a2, b1);
#endif // TRANSITION, GH-1635
```

File: tests/std/tests/P1614R2_spaceship/test.cpp [hunk 8]
```diff
@@ -232,61 +254,86 @@ void ordering_test_cases() {
...
+        ordered_iterator_test(a1.begin(), a1.begin(), a1.end(), a1.cbegin(), a1.cbegin(), a1.cend());
+    }
+    { // string_view
+        std::string_view a1 = "aaa";
+        std::string_view a2 = "aaa";
+        std::string_view b1 = "bb";
+        // ordered_containers_test(a1, a2, b1); // TRANSITION, GH-1635
```
Comment (CaseyCarter) [inline tests/std/tests/P1614R2_spaceship/test.cpp:309]:
```suggestion
#if 0 // TRANSITION, GH-1635
        std::string_view a2 = "aaa";
        std::string_view b1 = "bb";
        ordered_containers_test(a1, a2, b1);
#endif // TRANSITION, GH-1635
```

**Review by CaseyCarter — APPROVED** [review]
Apply warning cleanup for tests.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I've pushed the following changes:

* Fix `_HAS_CXX` comment typo.
* The `_Deque_const_iterator` spaceship shouldn't be `constexpr`.
* Mark `vector` and `vector<bool>` iterator spaceships as `_CONSTEXPR20_CONTAINER`.
* Fix `_STL_VERIFY` message for `_String_view_iterator` spaceship, which isn't equality.
* Skip `_Tree_const_iterator` `operator!=` for C++20.
* Implement spaceships for `stdext::checked_array_iterator` and `stdext::unchecked_array_iterator`.
* Skip `_Reinterpret_move_iter` `operator!=` for C++20. (This is an internal iterator used to make `unordered_meow` moves more efficient.)
* When testing `queue`, there's no need to test `deque` iterators again, as they were tested above.
* Activate, reorder, and comment "string iterators", "string_view iterators" tests.
* Test `checked_array_iterator` and `unchecked_array_iterator`.

File: stl/inc/vector [hunk 9]
```diff
@@ -153,12 +153,12 @@ public:
...
 
 #if _HAS_CXX20
-    _NODISCARD strong_ordering operator<=>(const _Vector_const_iterator& _Right) const noexcept {
+    _NODISCARD _CONSTEXPR20_CONTAINER strong_ordering operator<=>(const _Vector_const_iterator& _Right) const noexcept {
         _Compat(_Right);
         return _Unfancy(_Ptr) <=> _Unfancy(_Right._Ptr);
     }
 #else // ^^^ _HAS_CXX20 ^^^ / vvv !_HAS_CXX20 vvv
-    _NODISCARD bool operator!=(const _Vector_const_iterator& _Right) const noexcept {
+    _NODISCARD _CONSTEXPR20_CONTAINER bool operator!=(const _Vector_const_iterator& _Right) const noexcept {
```
Comment (CaseyCarter) [inline stl/inc/vector:161]: `_CONSTEXPR20_CONTAINER` is always simply `inline` in pre-C++20 modes. This is at best confusing - I think we should revert the change on this line. (Also on 2223.)

File: tests/std/tests/P1614R2_spaceship/test.cpp (hunk 4 above)
Comment (StephanTLavavej) [reply tests/std/tests/P1614R2_spaceship/test.cpp:192]: This exercises the constness scenarios so I believe they are valuable.

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/vector (hunk 9 above)
Comment (CaseyCarter) [reply stl/inc/vector:161]: STL points out that _all_ of these operators that are disabled in C++20 mode should not be `_CONSTEXPR20_CONTAINER`.

File: tests/std/tests/P1614R2_spaceship/test.cpp (hunk 4 above)
Comment (CaseyCarter) [reply tests/std/tests/P1614R2_spaceship/test.cpp:192]: Good enough for me.

File: stl/inc/vector (hunk 9 above)
Comment (StephanTLavavej) [reply stl/inc/vector]: Fixed all occurrences (I think) of `_CONSTEXPR20` and `_CONSTEXPR20_CONTAINER` appearing in `!_HAS_CXX20` mode.

There are occurrences in `<array>` and `<vector>` where `< <= > >=` (marked as `_CONSTEXPR20` and `_CONSTEXPR20_CONTAINER`) are guarded by `!defined(__cpp_lib_concepts)`; those cannot be changed yet.

`allocator` has an `operator!=` marked `_CONSTEXPR20_DYNALLOC`, but `allocator` is still on the todo list for Clause 20, not changed here.

**Review by CaseyCarter — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/vector (hunk 9 above)
Comment (CaseyCarter) [reply stl/inc/vector]: > There are occurrences in `<array>` and `<vector>` where `< <= > >=` (marked as `_CONSTEXPR20` and `_CONSTEXPR20_CONTAINER`) are guarded by `!defined(__cpp_lib_concepts)`; those cannot be changed yet.

Recorded [a cleanup reminder](https://github.com/microsoft/STL/issues/395#issuecomment-781809236) in #395.

**Comment by StephanTLavavej** [comment]
Thanks for implementing this Secret Bonus Round of spaceships! :1st_place_medal: :flying_saucer: :joy_cat:
